catalyst_voices/packages/catalyst_voices_services/lib/src/transaction/registration_transaction_builder.dart

```diff
@@ -0,0 +1,171 @@
+import 'dart:math';
+
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
+import 'package:catalyst_voices_models/catalyst_voices_models.dart';
+
+/// A builder that builds a Catalyst user registration transaction
+/// using RBAC specification.
+class RegistrationTransactionBuilder {
+  /// The RBAC registration purpose for the Catalyst Project.
+  static const _catalystUserRoleRegistrationPurpose =
+      'ca7a1457-ef9f-4c7f-9c74-7f8c4a4cfa6c';
+
+  /// The transaction config with current network parameters.
+  final TransactionBuilderConfig transactionConfig;
+
+  /// The network ID where the transaction will be submitted.
+  final NetworkId networkId;
+
+  /// The catalyst user seed phrase from which role specific keys are derived.
+  final SeedPhrase seedPhrase;
+
+  /// The user selected roles for which the user is registering.
+  final Set<AccountRole> roles;
+
+  /// The change address where the change from the transaction should go.
+  final ShelleyAddress changeAddress;
+
+  /// The list of user-related stake addresses.
+  ///
+  /// The first one will be used to as subjectAltName
+  /// in the certificate for the registration.
+  final List<ShelleyAddress> rewardAddresses;
+
+  /// The UTXOs that will be used as inputs for the transaction.
+  final Set<TransactionUnspentOutput> utxos;
+
+  const RegistrationTransactionBuilder({
+    required this.transactionConfig,
+    required this.networkId,
+    required this.seedPhrase,
+    required this.roles,
+    required this.changeAddress,
+    required this.rewardAddresses,
+    required this.utxos,
+  });
+
+  /// Builds the unsigned registration transaction.
+  Future<Transaction> build() async {
+    if (utxos.isEmpty) {
+      throw Exception('Insufficient balance, please top up your wallet');
+    }
+
+    final x509Envelope = await _buildMetadataEnvelope();
+
+    return _buildUnsignedRbacTx(
+      auxiliaryData: AuxiliaryData.fromCbor(
+        x509Envelope.toCbor(serializer: (e) => e.toCbor()),
+      ),
+    );
+  }
+
+  Future<X509MetadataEnvelope<RegistrationData>>
```

Comment:
feels like we should have some shorter typedef for it.

```dart
typedef XXX = X509MetadataEnvelope<RegistrationData>
```

Reply:
Came up with RegistrationMetadata